src/java.base/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -153,6 +154,51 @@ public static MethodHandle adaptUpcallForIMR(MethodHandle target, boolean dropRe
         return target;
     }
 
+    private static MethodType computeCollectArgumentsType(MethodType target, int insertIndex, MethodType collector) {
+        if (collector.returnType() != void.class){
+            // for non-void collector, the returned value is inserted into the argument list passed to the target
+            target = target.dropParameterTypes(insertIndex, insertIndex + 1);
+        }
+        return target.insertParameterTypes(insertIndex, collector.parameterArray()); // insert targetType parameters instead
+    }
+
+    // this just computes the adjusted type
+    private static MethodType computeUpcallIMRType(MethodType targetType, boolean dropReturn) {
```

Comment:
Crazy idea - what if we created a fake method handle of the given target type - then just called adaptUpcallForIMR on it, and got the resulting type? I'm a bit worried for the type logic and the MH adaptation logic to go out of sync, and require changing things in two places.

Reply:
Nice idea! This is trivial to implement as well using `MethodHandles::empty`.